### ISSUE TITLE
Elevation distance and illumination

### ISSUE THREAD

**Issue by gurgugnao** [Saibot393/perceptive#43]
Fantastic job with the distance penalty! Thank you very much once again. Being this module so useful I bother you with some more insights.

I just noticed that sight range, both spotting range and the new range modifier, does not consider elevation distance. The same happens with the illumination indicator. Moreover, illumination indicator only reacts to ambient light and not to lights emitted by tokens.
If it can be of any help, for illumination, the module Token Light Condition correctly applies the active effect depending on token elevation. Indeed, in that module seems that light is considered as a sphere, while Foundry normal behaviour is to consider lights as cylinder, if I am correct. In any case, independently of the token light condition, it is still visible to other tokens. I post some examples below.

1) Spotter token is at 15 feet from the hidden token. It shouldn't see it because of the sight range limit set to 10 feet. But it can because elevation distance is not calculated. Note that the spotter token has a dim light active effect even if it is inside the bright light radius (because it is elevated).
![Diapositiva1](https://github.com/Saibot393/perceptive/assets/122090327/f545c912-5c3f-4605-9103-8ae8547803f1)

2) Being elevated, the spotter token should be now in dark light (if the light is treated as a sphere) and has a dark light active effect. However, the illumination indicator is on dim light, as if it wasn't elevated.
![Diapositiva2](https://github.com/Saibot393/perceptive/assets/122090327/b14ec582-703b-45d3-bdde-17ee66ac0636)

3) As a result, the hidden token can see the spotter token even if it is in dark light (the hidden token is without darkvision).
![Diapositiva3](https://github.com/Saibot393/perceptive/assets/122090327/3e475a95-5be6-4fc2-80d3-bf9a1fd320b9)

**Comment by Saibot393** [comment]
I will look into it, thanks for the example screenshots, they are quite helpful. I will probably be able to add it with the update.

**Comment by gurgugnao** [comment]
This comment it is not important for the token elevation calculation, but maybe it can help you understanding the differences between Perceptive (and standard) illumination and Token Light Condition.

I noticed a different behaviour with proximity walls. In this case it is Token Light Condition that erroneously blocks light of walls (windows) set to "sight restriction: proximity 10 feet" (example 1 - the active effect on the token is dark light), while it behaves normally with (normal) windows set to "sight restriction: none" (example 2 - the active effect on the token is dim light). Perceptive illumination calculation is correct (dim) in both examples.

1) ![Diapositiva1](https://github.com/Saibot393/perceptive/assets/122090327/e7262d11-1012-46fc-90b3-a47914179c98)

2) ![Diapositiva2](https://github.com/Saibot393/perceptive/assets/122090327/5dcd388f-9b76-42de-b457-914316620943)

**Comment by Saibot393** [comment]
I have added a quick pathc with v2.8.2, token lights should now be included in the illumination calculation. The other things will take a bit longer.

**Comment by gurgugnao** [comment]
Of course, I just wanted to share with you my observations. Thank you for the prompt answer.

**Comment by Saibot393** [comment]
The new v2.9.0 has an option to calculate the illumination of tokens in 3D. This however only changes the Illumination used by Perceptive, that means that if a token should be invisible since it is outside of any light souce in 3D but not in 2D it will still show up. I have an idead how to add that , but that will take a bit longer to implement

**Comment by gurgugnao** [comment]
Nice!
Illumination calculation seems to work perfectly in 3D. On the other hand, sight range, both spotting range and the new range modifier, does not calculate 3D distance, just 2D.

**Comment by Saibot393** [comment]
I have added a new setting "Range 3D calculation" to activate 3D for vision with v2.9.4
